### ISSUE TITLE
Renku CLI crashes after generating new dataset files.

### ISSUE THREAD

**Issue by jkminder** [SwissDataScienceCenter/renku-python#3003]
## Details
I have updated two files that are contained in the same dataset with a python scripts (via renku run ...). Command was 

` renku run --name parliamentdb_conversion --input modules=conversion/democrasci_modules.py --input
            sqlite_iterator=conversion/sqlitedb_iterator.py --input sqlite_handler=conversion/sqlite_handler.py
            --input translations=conversion/resources/translations --input scripts=scripts --no-output-detection
            --output rdb=data/democrasci_parliamentdb/democrasci_parliamentdb_rdb.db --output
            graph=data/democrasci_parliamentdb/democrasci_parliamentdb_graph.dump -- python conversion/convert.py
            --schema conversion/democrasci_schema.yaml --constraints conversion/resources/unique_constraints.csv
            --delete --iterator data/parliamentdb/parliament.db --output
            data/democrasci_parliamentdb/democrasci_parliamentdb`

`git log` successfully prints the renku commit. But now some renku commands return the error below. I have tried `renku log` and `renku dataset update -a`, both fail with the same error. `renku dataset ls` still works normaly.

**Renku version:** 1.5.0
**OS:** Linux (#194-Ubuntu SMP Thu Jun 2 18:54:48 UTC 2022)
**Python:** 3.7.12

### Traceback

```
Traceback (most recent call last):
  File "[...]/site-packages/renku/ui/cli/exception_handler.py", line 134, in main
    return super().main(*args, **kwargs)
  File "[...]/site-packages/renku/ui/cli/exception_handler.py", line 93, in main
    return super().main(*args, **kwargs)
  File "[...]/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "[...]/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "[...]/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "[...]/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "[...]/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "[...]/site-packages/renku/ui/cli/dataset.py", line 1115, in update
    dry_run=dry_run,
  File "[...]/site-packages/renku/command/command_builder/command.py", line 296, in execute
    output = context["click_context"].invoke(self._operation, *args, **kwargs)
  File "[...]/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "[...]/site-packages/inject/__init__.py", line 342, in injection_wrapper
    return sync_func(*args, **kwargs)
  File "[...]/site-packages/renku/core/dataset/dataset.py", line 761, in update_datasets
    _update_datasets_files_metadata(client, updated_files, deleted_files, delete)
  File "[...]/site-packages/renku/core/dataset/dataset.py", line 1040, in _update_datasets_files_metadata
    datasets_provenance.add_or_update(dataset, creator=get_git_user(client.repository))
  File "[...]/site-packages/renku/core/dataset/datasets_provenance.py", line 110, in add_or_update
    self.dataset_gateway.add_or_remove(dataset)
  File "[...]/site-packages/renku/infrastructure/gateway/dataset_gateway.py", line 87, in add_or_remove
    database["datasets-provenance-tails"].pop(dataset.derived_from.url_id, None)
  File "[...]/site-packages/renku/infrastructure/database.py", line 661, in pop
    return self._entries.pop(key) if default is MARKER else self._entries.pop(key, default)
  File "[...]/site-packages/renku/infrastructure/database.py", line 434, in setstate
    data = self._storage.load(filename=self._get_filename_from_oid(object._p_oid))
  File "[...]/site-packages/renku/infrastructure/database.py", line 819, in load
    data = json.load(file)
  File "[...]/json/__init__.py", line 296, in load
    parse_constant=parse_constant, object_pairs_hook=object_pairs_hook, **kw)
  File "[...]/json/__init__.py", line 348, in loads
    return _default_decoder.decode(s)
  File "[...]/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "[...]/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 15 column 1 (char 455)
```

**Comment by Panaetius** [comment]
Could you tell me what project this is on? The error means some metadata is corrupt (could be due to a merge conflict), but without taking a look at the project, I couldn't tell the exact cause

**Comment by jkminder** [comment]
The project is private and hosted on renkulab.io. I have given you temporary access (hope @ralf.grubenmann was the correct handle). https://renkulab.io/gitlab/jminder/democrasci-parliamentdb/

**Comment by Panaetius** [comment]
If you check the file `.renku/metadata/datasets-provenance-tails` there is a leftover merge conflict that needs to be resolved.

You can just replace ```
<<<<<<< HEAD
                  "/datasets/566d7eaa108e4f459a3c0250868ce1b5",
                  {
                    "@renku_data_type": "renku.domain_model.dataset.Dataset",
                    "@renku_oid": "2cf8f96366e261c1ccedfb145543e1412257ef922d7be3547b3f602ee4d6cbb9",
                    "@renku_reference": true
                  },
                  "/datasets/b32a8df2b6214410b88b2cd3ca5d8a42",
                  {
                    "@renku_data_type": "renku.domain_model.dataset.Dataset",
                    "@renku_oid": "561610f551d7964ec306da66ad3f0273ab1037236d9956c459778d1c289567aa",
=======
                  "/datasets/28327d23f77948c1918b9d15303080b4",
                  {
                    "@renku_data_type": "renku.domain_model.dataset.Dataset",
                    "@renku_oid": "8d962e6d8cb8bab958e4e10713b8f7c13fc1ea63752694bd8aa8f11b388e94fe",
                    "@renku_reference": true
                  },
                  "/datasets/611ea88804df4dc9849a492d38ae1aec",
                  {
                    "@renku_data_type": "renku.domain_model.dataset.Dataset",
                    "@renku_oid": "89d83310cb7701795c3fd52f382f1e9ed900da844c9b4f829aa1c9a28262bd22",
>>>>>>> 0102e6475af64e71c85a28907867be36234aa64b
```
with
```
                  "/datasets/566d7eaa108e4f459a3c0250868ce1b5",
                  {
                    "@renku_data_type": "renku.domain_model.dataset.Dataset",
                    "@renku_oid": "2cf8f96366e261c1ccedfb145543e1412257ef922d7be3547b3f602ee4d6cbb9",
                    "@renku_reference": true
                  },
                  "/datasets/b32a8df2b6214410b88b2cd3ca5d8a42",
                  {
                    "@renku_data_type": "renku.domain_model.dataset.Dataset",
                    "@renku_oid": "561610f551d7964ec306da66ad3f0273ab1037236d9956c459778d1c289567aa",
                    "@renku_reference": true
                  },
                  "/datasets/28327d23f77948c1918b9d15303080b4",
                  {
                    "@renku_data_type": "renku.domain_model.dataset.Dataset",
                    "@renku_oid": "8d962e6d8cb8bab958e4e10713b8f7c13fc1ea63752694bd8aa8f11b388e94fe",
                    "@renku_reference": true
                  },
                  "/datasets/611ea88804df4dc9849a492d38ae1aec",
                  {
                    "@renku_data_type": "renku.domain_model.dataset.Dataset",
                    "@renku_oid": "89d83310cb7701795c3fd52f382f1e9ed900da844c9b4f829aa1c9a28262bd22",
```

Since version 1.3.0 we have a custom git merge tool (installed with `renku mergetool install`) that will automatically resolve all merge conflicts in renku metadata in most cases.

**Comment by Panaetius** [comment]
We don't force commit and we don't have a fail-then-fix approach, the mergetool gets called by git and handles merges of metadata on its own directly as part of a `git merge`. You don't call it manually and it doesn't fix these kinds of issues, as essentially the harm already happened when you get the `>>>>> HEAD` entries and the mergetool is there to prevent even getting to that point.

Looking through the repository history, I think here's what happened:
- You wanted to make a merge (without mergetool installed or at least not set up in the repo*) and got a metadata conflict.
- Git create the conflict file with `>>>>>> HEAD` in it and left the dirty 
- You then called `renku mergetool install`, which would set up the mergetool that would help with future merges (not the current, failed one). But that command looks like it has a bug, as it not only commits `.gitattributes` (that is has to commit) but also every other dirty file, and that's how you ended with this file committed to the code.

*git does not allow setting up a mergetool and have it be set up after clone, as configuration has to be in `.git/config` and that file does not get stored in git. So on every new clone, the mergetool has to bet set up again. We do that automatically when using `renku clone` for cloning, but we can't have a solution that ensures it's always installed and enabled due to the security limitations of git.

I've added a fix for that [here](https://github.com/SwissDataScienceCenter/renku-python/pull/3013) so this doesn't happen again.

**Comment by jkminder** [comment]
Aa okay I see I completely misunderstood the effect of the `renku mergetool install` command (`install` is perfectly clear, thats my fault). Now that I have an idea how this came to happen I have checked my recent commit history and I noticed that the issue was already there much longer, I just didn't notice. Upon closer examination I discovered that the insertion of the `>>>> HEAD` etc lines was happening together with the `renku mergetool install` renku commit. I don't know if the bug you were referring to above is related/the same but you might take a look at commit 262d21e64b0fd4fe892be3402b53fa22de11973e.

In any case, thank you for your continued help and quick support! I appreciate it.
